### PR TITLE
Fix/dealer

### PR DESCRIPTION
**Pull request by emielsteerneman**
### Comments for the reviewer
Fixed the bug where the dealer threw an assert error when there were more robots than roles. Added  robot.getPositionToDefend() to Dealer::getRobotScoreForDistance, although it's unsure yet if this is an improvement. 

###### Code Quality
- [x] Is the code is understandable and easy to read
- [x] Changes to the code comply with set clang-format rules
- [x] No use of manual memory control (e.g new/malloc/colloc etc)
- [x] Are (only) smart pointers used?

###### Testing
- [x] All tests are passing.
- [ ] I _added new / changed existing_ tests to reflect code changes (state why not otherwise!)
- [x] I tested my changes manually (Describe how, to what extent etc.)

I compared the role assignment to the dealer before this fix, and it's the same. I added print statements to manually check the cost matrix.

###### Commit Messages
- [x] Commit message is saying what has been changed, **why** it was changed? Remember other developers might not know
  what the problem you are fixing was. Note also negative _decision_ (e.g., why did you not do particular thing)
  **TLDR: Commit message are comprehensive**
- [x] Commit messages follows the rules of https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by emielsteerneman** [comment]
Fixed a ~segfault~ bad allocation exception in [distribute_forcedIDs](https://github.com/RoboTeamTwente/roboteam_ai/blob/818879115147ea7fdbdf3aa5c39e4c9910370f51/src/utilities/Dealer.cpp#L160). 

The function incorrectly assumed that all robots in the robot map were sorted by id (which is not guaranteed) and in the  range [0 .. n_robots] (which is incorrect if a robot is swapped out).

File: include/roboteam_ai/utilities/Dealer.h [hunk 1]
```diff
@@ -69,14 +70,14 @@ class Dealer {
...
-     * @param flagMap information map with all information for each role (with key -> roleName)
+     * Distributes the role between the friendly robots considering all information given in the role_map
+     * !!! robots and role_map are copies as they are modified to reduce computations after the forcedID's.
+     * @param robots vector containing all friendly robots
+     * @param role_map information map with all information for each role (with key -> roleName)
      * @param stpInfoMap
      * @return a vector with the roleName and the Robot that should get the role
      */
-    std::unordered_map<std::string, v::RobotView> distribute(std::vector<v::RobotView> allRobots, FlagMap flagMap, const std::unordered_map<std::string, stp::StpInfo> &stpInfoMap);
+    std::unordered_map<std::string, v::RobotView> distribute(std::vector<v::RobotView> robots, FlagMap role_map, const std::unordered_map<std::string, stp::StpInfo> &stpInfoMap);
```
Comment (Yuhanun) [inline include/roboteam_ai/utilities/Dealer.h]: Keep in mind that a robot view is still a reference to the original robot

Comment (emielsteerneman) [reply include/roboteam_ai/utilities/Dealer.h]: I'm not sure what you're saying. Do you mean the segfault danger is still there when the original robots disappear?

Comment (Yuhanun) [reply include/roboteam_ai/utilities/Dealer.h]: Yeah exactly, std::vector<robot> would be better, you can always construct a view from that

**Comment by Yuhanun** [comment]
LGTM

**Review by Yuhanun — APPROVED** [review]
LGTM

**Pull request merged**